dbcon/joblist/tuplehashjoin.cpp

```diff
@@ -200,140 +202,237 @@ void TupleHashJoinStep::join()
-            }
-
-            memUseAfter = joiner->getMemUsage() + rgDataSize;
+            utils::releaseSpinlock(rgdLock);
 
-            if (UNLIKELY(!flippedUMSwitch && (memUseAfter >= pmMemLimit)))
+            rgSize = smallRG.getSizeWithStrings();
+            atomicops::atomicAdd(&memUsedByEachJoin[index], rgSize);
+            gotMem = resourceManager->getMemory(rgSize, sessionMemLimit, false);
+            if (!gotMem)
             {
-                flippedUMSwitch = true;
-                oss << "UM join (" << index << ") ";
-#ifdef JLF_DEBUG
-                cout << oss.str() << endl;
-#endif
-                extendedInfo += oss.str();
-                joiner->setInUM();
-                memUseAfter = joiner->getMemUsage() + rgDataSize;
-            }
-
-            gotMem = resourceManager->getMemory(memUseAfter - memUseBefore, sessionMemLimit, false);
-            atomicops::atomicAdd(&totalUMMemoryUsage, memUseAfter - memUseBefore);
-            memUsedByEachJoin[index] += memUseAfter - memUseBefore;
-
-            /* This is kind of kludgy and overlaps with segreateJoiners() atm.
-            If this join won't be converted to disk-based, this fcn needs to abort the same as
-            it did before.  If it will be converted, it should just return. */
-            if (UNLIKELY(!gotMem))
-            {
-                if (isDML || !allowDJS || (fSessionId & 0x80000000) ||
-                        (tableOid() < 3000 && tableOid() >= 1000))
+                boost::unique_lock<boost::mutex> sl(saneErrMsg);
+                if (!joinIsTooBig && (isDML || !allowDJS || (fSessionId & 0x80000000) ||
+                    (tableOid() < 3000 && tableOid() >= 1000)))
                 {
                     joinIsTooBig = true;
                     fLogger->logMessage(logging::LOG_TYPE_INFO, logging::ERR_JOIN_TOO_BIG);
                     errorMessage(logging::IDBErrorInfo::instance()->errorMsg(logging::ERR_JOIN_TOO_BIG));
                     status(logging::ERR_JOIN_TOO_BIG);
-                    cout << "Join is too big, raise the UM join limit for now" << endl;
+                    cout << "Join is too big, raise the UM join limit for now (small runner)" << endl;
                     abort();
                     break;
                 }
                 else
+                {
+                    joiner->setConvertToDiskJoin();
                     return;
+                }
             }
 
+            joiner->insertRGData(smallRG, threadID);
+
+            if (!joiner->inUM() && (memUsedByEachJoin[index] > pmMemLimit))
+            {
+                joiner->setInUM(rgData[index]);
+                for (int i = 1; i < numCores; i++)
+                    jobs[i] = jobstepThreadPool.invoke([this, i, index, jobs]
+                        { this->smallRunnerFcn(index, i, jobs); });
+            }
 next:
-//  		cout << "inserted one rg into the joiner, rowcount = " <<
-//  			smallRG.getRowCount() << endl;
+            dlMutex.lock();
```

Comment:
That is what you were referering to comparing consequences of internal and external locking for FIFO.

Reply:
Why do you need to lock given that smallDL is a dedicated FIFO serving the thread?

Reply:
There are potentially numCores threads running this code at once. 

Yeah that's what I was talking about.  fifo::next() doesn't use its lock in such a way that multiple threads could use the same iterator at once -> need to lock externally to do that.